### PR TITLE
 Webhook e2e test: fail open and fail closed

### PR DESCRIPTION
**Pull request by janetkuo**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Ref: https://github.com/kubernetes/features/issues/492

**Special notes for your reviewer**: ~depends on #55127~ (merged) 
@kubernetes/sig-api-machinery-api-reviews

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/.golint_failures [hunk 1]
```diff
@@ -551,6 +551,8 @@ staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1
 staging/src/k8s.io/apiserver/pkg/apis/audit/validation
 staging/src/k8s.io/apiserver/pkg/apis/example
 staging/src/k8s.io/apiserver/pkg/apis/example/v1
```
Comment (liggitt) [inline hack/.golint_failures]: @sttts @deads2k why are we shipping fake api types inside k8s.io/apiserver? I thought we had https://github.com/kubernetes/sample-apiserver/tree/master/pkg/apis/wardle for that. Multiple fake api groups/types inside a server library doesn't seem like a good idea.

File: staging/src/k8s.io/apiserver/pkg/apis/example2/v1/types.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ReplicaSet ensures that a specified number of pod replicas are running at any given time.
+type ReplicaSet struct {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/example2/v1/types.go]: if we need a fake type for this test, why not use a CRD? can define it just for the test, and the webhook can expect a specific GVK

Comment (janetkuo) [reply staging/src/k8s.io/apiserver/pkg/apis/example2/v1/types.go]: These changes are from #55127, cc @caesarxuchao

Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/apis/example2/v1/types.go]: We have other 2 tests for CRD.

This fake type is to test cross group conversion. The internal type of this "ReplicaSet" is registered in the "example" group, and the external types if registered in the "example2" group. @deads2k requested adding such a test, and I think it's valuable.

File: test/e2e/apimachinery/webhook.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ import (
 	crdclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	"k8s.io/apiextensions-apiserver/test/integration/testserver"
 	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
Comment (mml) [inline test/e2e/apimachinery/webhook.go]: This looks like the same package imported with two names?

File: test/e2e/apimachinery/webhook.go [hunk 4]
```diff
@@ -256,6 +257,10 @@ func registerWebhook(f *framework.Framework, context *certContext) {
 	By("Registering the webhook via the AdmissionRegistration API")
 
 	namespace := f.Namespace.Name
+	// Create a webhook that always fails (fail open)
```
Comment (mml) [inline test/e2e/apimachinery/webhook.go]: We use "always fails" to describe both "fail open" and "fail closed".  (Other two marked below).  I don't actually understand what this test is trying to test.  Maybe we can add a short comment that explain the overall test plan and behavior that we're trying to verify?

File: test/e2e/apimachinery/webhook.go [hunk 5]
```diff
@@ -309,6 +314,8 @@ func registerWebhook(f *framework.Framework, context *certContext) {
 					CABundle: context.signingCert,
 				},
 			},
+			// Calling this webhook will always fail (fail open)
```
Comment (mml) [inline test/e2e/apimachinery/webhook.go]: (always fails)

File: test/e2e/apimachinery/webhook.go [hunk 6]
```diff
@@ -367,6 +374,65 @@ func testWebhook(f *framework.Framework) {
...
+	namespace := f.Namespace.Name
+	policyFail := v1alpha1.Fail
+	hook := failingWebhook(namespace, "fail-closed.k8s.io")
+	hook.FailurePolicy = &policyFail
+	_, err := client.AdmissionregistrationV1alpha1().ValidatingWebhookConfigurations().Create(&v1alpha1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: webhookFailClosedConfigName,
+		},
+		Webhooks: []v1alpha1.Webhook{
+			// Calling this webhook will always fail (fail closed)
```
Comment (mml) [inline test/e2e/apimachinery/webhook.go]: (always fails)

**Review by mml — CHANGES_REQUESTED** [review]

**Comment by janetkuo** [comment]
@mml comments addressed, PTAL

**Comment by janetkuo** [comment]
Rebased

**Comment by janetkuo** [comment]
Added another commit to use namespace selector in the fail-closed webhook. Otherwise, other AdmissionWebhook e2e tests (when running in parallel) may be affected, because this fail-closed webhook rejects configmap creation.

File: hack/.golint_failures (hunk 1 above)
Comment (sttts) [reply hack/.golint_failures]: @liggitt we cannot reference sample-apiserver from apiserver. And we need some types to test. They are really only fake, stripped down and no need to keep in sync with the real core types.

**Comment by janetkuo** [comment]
Rebased

File: hack/.golint_failures (hunk 1 above)
Comment (liggitt) [reply hack/.golint_failures]: ok, looked real enough to confuse/concern me

**Review by mml — APPROVED** [review]

**Comment by mml** [comment]
Thanks, @janetkuo.  Makes more sense now.

/lgtm

**Comment by janetkuo** [comment]
Squashed commits

**Pull request merged**